File: resources/android/TextAreaControl.kt
package dev.firstlightui.plugins.firstlight_ui.ui

import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.error
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.input.TextFieldValue
import androidx.compose.ui.unit.dp
import com.nativephp.plugins.native_ui.NativeUITokens

@Composable
internal fun FirstlightTextAreaControl(
    configuration: TextAreaRendererConfiguration,
    draft: TextFieldValue,
    tokens: NativeUITokens,
    onValueChange: (TextFieldValue) -> Unit,
    modifier: Modifier = Modifier,
) {
    val supporting = configuration.error.ifEmpty { configuration.helper }
    val description = listOf(configuration.accessibilityLabel, configuration.accessibilityHint)
        .filter(String::isNotEmpty)
        .joinToString(". ")

    OutlinedTextField(
        value = draft,
        onValueChange = onValueChange,
        modifier = modifier
            .fillMaxWidth()
            .defaultMinSize(minHeight = 48.dp)
            .semantics {
                if (description.isNotEmpty()) contentDescription = description
                if (configuration.error.isNotEmpty()) error(configuration.error)
            },
        enabled = !configuration.disabled,
        readOnly = configuration.readOnly,
        singleLine = false,
        minLines = configuration.minLines,
        maxLines = configuration.maxLines,
        isError = configuration.error.isNotEmpty(),
        label = configuration.label.takeIf(String::isNotEmpty)?.let { label ->
            { Text(if (configuration.required) "$label *" else label) }
        },
        placeholder = configuration.placeholder.takeIf(String::isNotEmpty)?.let { placeholder ->
            { Text(placeholder) }
        },
        supportingText = supporting.takeIf(String::isNotEmpty)?.let { text ->
            {
                Text(
                    text = text,
                    color = if (configuration.error.isEmpty()) tokens.onSurfaceVariant else tokens.destructive,
                )
            }
        },
        keyboardOptions = textAreaKeyboardOptions(configuration),
    )
}

File: resources/ios/TextAreaControl.swift
import SwiftUI
import UIKit

struct FirstlightTextAreaControl: View {
    let configuration: TextAreaRendererConfiguration
    @Binding var text: String
    let isFocused: FocusState<Bool>.Binding
    let tokens: NativeUITokens

    var minimumHeight: CGFloat {
        CGFloat(configuration.minLines) * 22 + 18
    }

    var maximumHeight: CGFloat {
        CGFloat(configuration.maxLines) * 22 + 18
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            if !configuration.label.isEmpty {
                Text(configuration.required ? "\(configuration.label) *" : configuration.label)
                    .font(.subheadline.weight(.medium))
                    .foregroundStyle(tokens.onSurface)
                    .fixedSize(horizontal: false, vertical: true)
            }

            ZStack(alignment: .topLeading) {
                if text.isEmpty && !configuration.placeholder.isEmpty {
                    Text(configuration.placeholder)
                        .foregroundStyle(Color(uiColor: .placeholderText))
                        .padding(.horizontal, 5)
                        .padding(.vertical, 8)
                        .allowsHitTesting(false)
                        .accessibilityHidden(true)
                }

                TextEditor(text: configuration.readOnly ? .constant(text) : $text)
                    .scrollContentBackground(.hidden)
                    .focused(isFocused)
                    .disabled(configuration.disabled)
                    .modifier(TextAreaCapitalizationModifier(policy: configuration.autocapitalize))
                    .modifier(TextAreaAutocorrectionModifier(policy: configuration.autocorrectPolicy))
                    .modifier(TextAreaAccessibilityLabelModifier(label: configuration.accessibilityLabel))
                    .modifier(TextAreaAccessibilityHintModifier(hint: configuration.accessibilityHint))
                    .accessibilityValue(TextAreaAccessibility.value(text: text, error: configuration.error))
            }
            .frame(minHeight: minimumHeight, maxHeight: maximumHeight, alignment: .topLeading)
            .padding(.horizontal, 7)
            .background(Color(uiColor: .secondarySystemBackground))
            .clipShape(RoundedRectangle(cornerRadius: 10, style: .continuous))
            .overlay {
                RoundedRectangle(cornerRadius: 10, style: .continuous)
                    .stroke(
                        configuration.error.isEmpty ? Color(uiColor: .separator) : tokens.destructive,
                        lineWidth: configuration.error.isEmpty ? 0.5 : 1,
                    )
            }

            let supporting = configuration.error.isEmpty ? configuration.helper : configuration.error
            if !supporting.isEmpty {
                Text(supporting)
                    .font(.footnote)
                    .foregroundStyle(configuration.error.isEmpty ? tokens.onSurfaceVariant : tokens.destructive)
                    .fixedSize(horizontal: false, vertical: true)
                    .accessibilityLabel(configuration.error.isEmpty ? supporting : "Error: \(supporting)")
            }
        }
        .opacity(configuration.disabled ? 0.6 : 1)
    }
}

enum TextAreaAccessibility {
    static func value(text: String, error: String) -> String {
        guard !error.isEmpty else { return text }
        guard !text.isEmpty else { return "Error: \(error)" }
        return "\(text). Error: \(error)"
    }
}

private struct TextAreaCapitalizationModifier: ViewModifier {
    let policy: String

    @ViewBuilder
    func body(content: Content) -> some View {
        switch policy {
        case "none": content.textInputAutocapitalization(.never)
        case "sentences": content.textInputAutocapitalization(.sentences)
        case "words": content.textInputAutocapitalization(.words)
        case "characters": content.textInputAutocapitalization(.characters)
        default: content
        }
    }
}

private struct TextAreaAutocorrectionModifier: ViewModifier {
    let policy: String

    @ViewBuilder
    func body(content: Content) -> some View {
        switch policy {
        case "enabled": content.autocorrectionDisabled(false)
        case "disabled": content.autocorrectionDisabled(true)
        default: content
        }
    }
}

private struct TextAreaAccessibilityLabelModifier: ViewModifier {
    let label: String

    @ViewBuilder
    func body(content: Content) -> some View {
        if label.isEmpty { content } else { content.accessibilityLabel(label) }
    }
}

private struct TextAreaAccessibilityHintModifier: ViewModifier {
    let hint: String

    @ViewBuilder
    func body(content: Content) -> some View {
        if hint.isEmpty { content } else { content.accessibilityHint(hint) }
    }
}
